File: HabboHotel/Rooms/Trade.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Uber.HabboHotel.GameClients;
using Uber.HabboHotel.Items;
using Uber.Messages;

namespace Uber.HabboHotel.Rooms
{
    class Trade
    {
        private ConcurrentDictionary<uint, TradeUser> Users;
        private int TradeStage;
        private uint RoomId;

        private uint oneId;
        private uint twoId;

        public Trade(uint UserOneId, uint UserTwoId, uint RoomId)
        {
            this.oneId = UserOneId;
            this.twoId = UserTwoId;

            this.Users = new ConcurrentDictionary<uint, TradeUser>();
            this.Users.TryAdd(UserOneId, new TradeUser(UserOneId, RoomId));
            this.Users.TryAdd(UserTwoId, new TradeUser(UserTwoId, RoomId));
            this.TradeStage = 1;
            this.RoomId = RoomId;

            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                if (!User.GetRoomUser().Statusses.ContainsKey("trd"))
                {
                    User.GetRoomUser().AddStatus("trd", "");
                    User.GetRoomUser().UpdateNeeded = true;
                }
            }

            ServerMessage Message = new ServerMessage(104);
            Message.AppendUInt(UserOneId);
            Message.AppendBoolean(true);
            Message.AppendUInt(UserTwoId);
            Message.AppendBoolean(true);
            SendMessageToUsers(Message);
        }

        public bool AllUsersAccepted
        {
            get
            {
                foreach (var kvp in Users)
                {
                    TradeUser User = kvp.Value;

                    if (!User.HasAccepted)
                        {
                            return false;
                        }
                    }

                return true;
            }
        }

        public bool ContainsUser(uint Id)
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                if (User.UserId == Id)
                    {
                        return true;
                    }
                }

            return false;
        }

        public TradeUser GetTradeUser(uint Id)
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                if (User.UserId == Id)
                    {
                        return User;
                    }
                }

            return null;
        }

        public void OfferItem(uint UserId, UserItem Item)
        {
            TradeUser User = GetTradeUser(UserId);

            if (User == null || Item == null || !Item.GetBaseItem().AllowTrade || User.HasAccepted || TradeStage != 1)
            {
                return;
            }

            ClearAccepted();

            User.OfferedItems.TryAdd(Item.Id, Item);
            UpdateTradeWindow();
        }

        public void TakeBackItem(uint UserId, UserItem Item)
        {
            TradeUser User = GetTradeUser(UserId);

            if (User == null || Item == null || User.HasAccepted || TradeStage != 1)
            {
                return;
            }

            ClearAccepted();

            User.OfferedItems.TryRemove(Item.Id, out var _);
            UpdateTradeWindow();
        }

        public void Accept(uint UserId)
        {
            TradeUser User = GetTradeUser(UserId);

            if (User == null || TradeStage != 1)
            {
                return;
            }

            User.HasAccepted = true;

            ServerMessage Message = new ServerMessage(109);
            Message.AppendUInt(UserId);
            Message.AppendBoolean(true);
            SendMessageToUsers(Message);

            if (AllUsersAccepted)
            {
                SendMessageToUsers(new ServerMessage(111));
                TradeStage++;
                ClearAccepted();
            }
        }

        public void Unaccept(uint UserId)
        {
            TradeUser User = GetTradeUser(UserId);

            if (User == null || TradeStage != 1 || AllUsersAccepted)
            {
                return;
            }

            User.HasAccepted = false;

            ServerMessage Message = new ServerMessage(109);
            Message.AppendUInt(UserId);
            Message.AppendBoolean(false);
            SendMessageToUsers(Message);
        }

        public void CompleteTrade(uint UserId)
        {
            TradeUser User = GetTradeUser(UserId);

            if (User == null || TradeStage != 2)
            {
                return;
            }

            User.HasAccepted = true;

            ServerMessage Message = new ServerMessage(109);
            Message.AppendUInt(UserId);
            Message.AppendBoolean(true);
            SendMessageToUsers(Message);

            if (AllUsersAccepted)
            {
                TradeStage = 999;
                DeliverItems();
                CloseTradeClean();
            }
        }

        public void ClearAccepted()
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                User.HasAccepted = false;
            }
        }

        public void UpdateTradeWindow()
        {
            ServerMessage Message = new ServerMessage(108);

            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                Message.AppendUInt(User.UserId);
                Message.AppendInt32(User.OfferedItems.Count);

                foreach (var offeredItems in User.OfferedItems)
                {
                    UserItem Item = offeredItems.Value;

                    Message.AppendUInt(Item.Id);
                    Message.AppendStringWithBreak(Item.GetBaseItem().Type.ToLower());
                    Message.AppendUInt(Item.Id);
                    Message.AppendInt32(Item.GetBaseItem().SpriteId);
                    Message.AppendBoolean(true);
                    Message.AppendBoolean(true);
                    Message.AppendStringWithBreak("");
                    Message.AppendBoolean(false); // xmas 09 furni had a special furni tag here, with wired day (wat?)
                    Message.AppendBoolean(false); // xmas 09 furni had a special furni tag here, wired month (wat?)
                    Message.AppendBoolean(false); // xmas 09 furni had a special furni tag here, wired year (wat?)

                    if (Item.GetBaseItem().Type.ToLower() == "s")
                    {
                        Message.AppendInt32(-1);
                    }
                }
            }

            SendMessageToUsers(Message);
        }

        public void DeliverItems()
        {
            // List items
            ConcurrentDictionary<uint, UserItem> ItemsOne = GetTradeUser(oneId).OfferedItems;
            ConcurrentDictionary<uint, UserItem> ItemsTwo = GetTradeUser(twoId).OfferedItems;

            // Verify they are still in user inventory
            foreach (var kvp in ItemsOne)
            {
                UserItem I = kvp.Value;

                if (GetTradeUser(oneId).GetClient().GetHabbo().GetInventoryComponent().GetItem(I.Id) == null)
                {
                    GetTradeUser(oneId).GetClient().SendNotif("Trade failed.");
                    GetTradeUser(twoId).GetClient().SendNotif("Trade failed.");

                    return;
                }
            }

            foreach (var kvp in ItemsTwo)
            {
                UserItem I = kvp.Value;

                if (GetTradeUser(twoId).GetClient().GetHabbo().GetInventoryComponent().GetItem(I.Id) == null)
                {
                    GetTradeUser(oneId).GetClient().SendNotif("Trade failed.");
                    GetTradeUser(twoId).GetClient().SendNotif("Trade failed.");

                    return;
                }
            }

            // Deliver them
            foreach (var kvp in ItemsOne)
            {
                UserItem I = kvp.Value;

                GetTradeUser(oneId).GetClient().GetHabbo().GetInventoryComponent().RemoveItem(I.Id);
                GetTradeUser(twoId).GetClient().GetHabbo().GetInventoryComponent().AddItem(I.Id, I.BaseItem, I.ExtraData);
            }

            foreach (var kvp in ItemsTwo)
            {
                UserItem I = kvp.Value;

                GetTradeUser(twoId).GetClient().GetHabbo().GetInventoryComponent().RemoveItem(I.Id);
                GetTradeUser(oneId).GetClient().GetHabbo().GetInventoryComponent().AddItem(I.Id, I.BaseItem, I.ExtraData);
            }

            // Update inventories
            GetTradeUser(oneId).GetClient().GetHabbo().GetInventoryComponent().UpdateItems(false);
            GetTradeUser(twoId).GetClient().GetHabbo().GetInventoryComponent().UpdateItems(false);
        }

        public void CloseTradeClean()
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                User.GetRoomUser().RemoveStatus("trd");
                    User.GetRoomUser().UpdateNeeded = true;
                }

            SendMessageToUsers(new ServerMessage(112));
            GetRoom().ActiveTrades.Remove(this);
        }

        public void CloseTrade(uint UserId)
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                if (User.GetRoomUser() == null)
                    {
                        continue;
                    }

                    User.GetRoomUser().RemoveStatus("trd");
                    User.GetRoomUser().UpdateNeeded = true;
                }

            ServerMessage Message = new ServerMessage(110);
            Message.AppendUInt(UserId);
            SendMessageToUsers(Message);
        }

        public void SendMessageToUsers(ServerMessage Message)
        {
            foreach (var kvp in Users)
            {
                TradeUser User = kvp.Value;

                User.GetClient().SendMessage(Message);
            }
        }

        private Room GetRoom()
        {
            return UberEnvironment.GetGame().GetRoomManager().GetRoom(RoomId);
        }
    }

    class TradeUser
    {
        public uint UserId;
        private uint RoomId;
        private bool Accepted;

        public ConcurrentDictionary<uint, UserItem> OfferedItems;

        public bool HasAccepted
        {
            get
            {
                return Accepted;
            }

            set
            {
                Accepted = value;
            }
        }

        public TradeUser(uint UserId, uint RoomId)
        {
            this.UserId = UserId;
            this.RoomId = RoomId;
            this.Accepted = false;
            this.OfferedItems = new ConcurrentDictionary<uint, UserItem>();
        }

        public RoomUser GetRoomUser()
        {
            Room Room = UberEnvironment.GetGame().GetRoomManager().GetRoom(RoomId);

            if (Room == null)
            {
                return null;
            }
            
            return Room.GetRoomUserByHabbo(UserId);
        }

        public GameClient GetClient()
        {
            return UberEnvironment.GetGame().GetClientManager().GetClientByHabbo(UserId);
        }
    }
}

File: src/main/java/com/uber/server/rooms/Trade.java
package com.uber.server.rooms;

import com.uber.server.game.Game;
import com.uber.server.game.GameClient;
import com.uber.server.items.Item;
import com.uber.server.messages.ServerMessage;
import com.uber.server.users.inventory.UserItem;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Represents a trade between two users.
 * Ported from HabboHotel/Rooms/Trade.cs
 */
public class Trade {
    private static final Logger logger = LoggerFactory.getLogger(Trade.class);
    
    private final long oneId;
    private final long twoId;
    private final long roomId;
    private final Game game;
    private final ConcurrentHashMap<Long, TradeUser> users;
    private int tradeStage;
    
    public Trade(long userOneId, long userTwoId, long roomId, Game game) {
        this.oneId = userOneId;
        this.twoId = userTwoId;
        this.roomId = roomId;
        this.game = game;
        this.users = new ConcurrentHashMap<>();
        this.tradeStage = 1;
        
        users.put(userOneId, new TradeUser(userOneId, roomId, game));
        users.put(userTwoId, new TradeUser(userTwoId, roomId, game));
        
        // Add trade status to room users
        Room room = getRoom();
        if (room != null) {
            for (TradeUser tradeUser : users.values()) {
                RoomUser roomUser = tradeUser.getRoomUser();
                if (roomUser != null && !roomUser.getStatuses().containsKey("trd")) {
                    roomUser.addStatus("trd", "");
                    roomUser.setUpdateNeeded(true);
                }
            }
        }
        
        // Send trade start message
        ServerMessage message = new ServerMessage(104);
        message.appendUInt(userOneId);
        message.appendBoolean(true);
        message.appendUInt(userTwoId);
        message.appendBoolean(true);
        sendMessageToUsers(message);
    }
    
    /**
     * Checks if all users have accepted the trade.
     * @return True if all users accepted
     */
    public boolean isAllUsersAccepted() {
        for (TradeUser user : users.values()) {
            if (!user.hasAccepted()) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Checks if trade contains a user.
     * @param userId User ID
     * @return True if user is in trade
     */
    public boolean containsUser(long userId) {
        return users.containsKey(userId);
    }
    
    /**
     * Gets a TradeUser by ID.
     * @param userId User ID
     * @return TradeUser object, or null if not found
     */
    public TradeUser getTradeUser(long userId) {
        return users.get(userId);
    }
    
    /**
     * Offers an item to the trade.
     * Ported from Trade.cs OfferItem()
     * @param userId User ID
     * @param item UserItem to offer
     */
    public void offerItem(long userId, UserItem item) {
        TradeUser tradeUser = getTradeUser(userId);
        if (tradeUser == null || item == null || tradeUser.hasAccepted() || tradeStage != 1) {
            return;
        }
        
        // Check if item allows trade
        Item baseItem = item.getBaseItem();
        if (baseItem == null || !baseItem.allowTrade()) {
            return;
        }
        
        clearAccepted();
        tradeUser.addOfferedItem(item);
        updateTradeWindow();
    }
    
    /**
     * Takes back an item from the trade.
     * Ported from Trade.cs TakeBackItem()
     * @param userId User ID
     * @param item UserItem to take back
     */
    public void takeBackItem(long userId, UserItem item) {
        TradeUser tradeUser = getTradeUser(userId);
        if (tradeUser == null || item == null || tradeUser.hasAccepted() || tradeStage != 1) {
            return;
        }
        
        clearAccepted();
        tradeUser.removeOfferedItem(item.getId());
        updateTradeWindow();
    }
    
    /**
     * Accepts the trade (stage 1).
     * Ported from Trade.cs Accept()
     * @param userId User ID
     */
    public void accept(long userId) {
        TradeUser tradeUser = getTradeUser(userId);
        if (tradeUser == null || tradeStage != 1) {
            return;
        }
        
        tradeUser.setAccepted(true);
        
        ServerMessage message = new ServerMessage(109);
        message.appendUInt(userId);
        message.appendBoolean(true);
        sendMessageToUsers(message);
        
        if (isAllUsersAccepted()) {
            sendMessageToUsers(new ServerMessage(111));
            tradeStage = 2;
            clearAccepted();
        }
    }
    
    /**
     * Unaccepts the trade (stage 1).
     * Ported from Trade.cs Unaccept()
     * @param userId User ID
     */
    public void unaccept(long userId) {
        TradeUser tradeUser = getTradeUser(userId);
        if (tradeUser == null || tradeStage != 1 || isAllUsersAccepted()) {
            return;
        }
        
        tradeUser.setAccepted(false);
        
        ServerMessage message = new ServerMessage(109);
        message.appendUInt(userId);
        message.appendBoolean(false);
        sendMessageToUsers(message);
    }
    
    /**
     * Completes the trade (stage 2).
     * Ported from Trade.cs CompleteTrade()
     * @param userId User ID
     */
    public void completeTrade(long userId) {
        TradeUser tradeUser = getTradeUser(userId);
        if (tradeUser == null || tradeStage != 2) {
            return;
        }
        
        tradeUser.setAccepted(true);
        
        ServerMessage message = new ServerMessage(109);
        message.appendUInt(userId);
        message.appendBoolean(true);
        sendMessageToUsers(message);
        
        if (isAllUsersAccepted()) {
            tradeStage = 999;
            deliverItems();
            closeTradeClean();
        }
    }
    
    /**
     * Clears accepted status for all users.
     * Ported from Trade.cs ClearAccepted()
     */
    public void clearAccepted() {
        for (TradeUser user : users.values()) {
            user.setAccepted(false);
        }
    }
    
    /**
     * Updates the trade window.
     * Ported from Trade.cs UpdateTradeWindow()
     */
    public void updateTradeWindow() {
        ServerMessage message = new ServerMessage(108);
        
        for (TradeUser tradeUser : users.values()) {
            message.appendUInt(tradeUser.getUserId());
            List<UserItem> offeredItems = tradeUser.getOfferedItems();
            message.appendInt32(offeredItems.size());
            
            for (UserItem item : offeredItems) {
                Item baseItem = item.getBaseItem();
                if (baseItem == null) {
                    continue;
                }
                
                message.appendUInt(item.getId());
                message.appendStringWithBreak(baseItem.getType().toLowerCase());
                message.appendUInt(item.getId());
                message.appendInt32(baseItem.getSpriteId());
                message.appendBoolean(true);
                message.appendBoolean(true);
                message.appendStringWithBreak(item.getExtraData() != null ? item.getExtraData() : "");
                message.appendBoolean(false); // xmas 09 special tag
                message.appendBoolean(false); // xmas 09 special tag
                message.appendBoolean(false); // xmas 09 special tag
                
                if ("s".equalsIgnoreCase(baseItem.getType())) {
                    message.appendInt32(-1);
                }
            }
        }
        
        sendMessageToUsers(message);
    }
    
    /**
     * Delivers items to users.
     * Ported from Trade.cs DeliverItems()
     */
    public void deliverItems() {
        TradeUser userOne = getTradeUser(oneId);
        TradeUser userTwo = getTradeUser(twoId);
        
        if (userOne == null || userTwo == null) {
            return;
        }
        
        GameClient clientOne = userOne.getClient();
        GameClient clientTwo = userTwo.getClient();
        
        if (clientOne == null || clientTwo == null || 
            clientOne.getHabbo() == null || clientTwo.getHabbo() == null) {
            return;
        }
        
        // Verify items are still in inventory
        List<UserItem> itemsOne = userOne.getOfferedItems();
        List<UserItem> itemsTwo = userTwo.getOfferedItems();
        
        for (UserItem item : itemsOne) {
            if (clientOne.getHabbo().getInventoryComponent().getItem(item.getId()) == null) {
                clientOne.sendNotif("Trade failed.");
                clientTwo.sendNotif("Trade failed.");
                return;
            }
        }
        
        for (UserItem item : itemsTwo) {
            if (clientTwo.getHabbo().getInventoryComponent().getItem(item.getId()) == null) {
                clientOne.sendNotif("Trade failed.");
                clientTwo.sendNotif("Trade failed.");
                return;
            }
        }
        
        // Deliver items
        for (UserItem item : itemsOne) {
            clientOne.getHabbo().getInventoryComponent().removeItem(item.getId());
            clientTwo.getHabbo().getInventoryComponent().addItem(item.getId(), item.getBaseItemId(), item.getExtraData());
        }
        
        for (UserItem item : itemsTwo) {
            clientTwo.getHabbo().getInventoryComponent().removeItem(item.getId());
            clientOne.getHabbo().getInventoryComponent().addItem(item.getId(), item.getBaseItemId(), item.getExtraData());
        }
        
        // Update inventories
        clientOne.getHabbo().getInventoryComponent().updateItems(false);
        clientTwo.getHabbo().getInventoryComponent().updateItems(false);
    }
    
    /**
     * Closes trade cleanly (after completion).
     * Ported from Trade.cs CloseTradeClean()
     */
    public void closeTradeClean() {
        for (TradeUser tradeUser : users.values()) {
            RoomUser roomUser = tradeUser.getRoomUser();
            if (roomUser != null) {
                roomUser.removeStatus("trd");
                roomUser.setUpdateNeeded(true);
            }
        }
        
        sendMessageToUsers(new ServerMessage(112));
        
        // Remove from room's active trades
        Room room = getRoom();
        if (room != null) {
            room.removeActiveTrade(this);
        }
    }
    
    /**
     * Closes trade (cancelled).
     * Ported from Trade.cs CloseTrade()
     * @param userId User ID who closed the trade
     */
    public void closeTrade(long userId) {
        for (TradeUser tradeUser : users.values()) {
            RoomUser roomUser = tradeUser.getRoomUser();
            if (roomUser != null) {
                roomUser.removeStatus("trd");
                roomUser.setUpdateNeeded(true);
            }
        }
        
        ServerMessage message = new ServerMessage(110);
        message.appendUInt(userId);
        sendMessageToUsers(message);
    }
    
    /**
     * Sends a message to both trade users.
     * Ported from Trade.cs SendMessageToUsers()
     * @param message ServerMessage to send
     */
    public void sendMessageToUsers(ServerMessage message) {
        if (message == null) {
            return;
        }
        
        for (TradeUser tradeUser : users.values()) {
            GameClient client = tradeUser.getClient();
            if (client != null) {
                client.sendMessage(message);
            }
        }
    }
    
    /**
     * Gets the Room this trade is in.
     * @return Room object, or null if not found
     */
    private Room getRoom() {
        if (game == null || game.getRoomManager() == null) {
            return null;
        }
        return game.getRoomManager().getRoom(roomId);
    }
    
    public long getOneId() { return oneId; }
    public long getTwoId() { return twoId; }
    public long getRoomId() { return roomId; }
    public int getTradeStage() { return tradeStage; }
}
